### ISSUE TITLE
Significant performance loss on openid-connect/token from 18.0.0 to 20.0.3

### ISSUE THREAD

**Issue by jhollmannk** [keycloak/keycloak#16909]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

core

### Describe the bug

Getting a token via openid-connect/token for a user which is in a group with a decent amount of clients having composite roles takes a really long time in 20.0.3. It was way faster in 18.0.0.

Duration in 18.0.0 is something around 100ms. Duration in 20.0.3 is over 6500ms.

These durations are taken for a user which is in a group with 400 clients each having 15 composite roles containing 2 roles. The client_id for the login via openid-connect/token is *not* associated with the group, so the resulting access_token has no resource_access entries at all!

That is puzzling me. Why does it take so much time to determine that the access_token is just empty?

Getting the group representation (with all the associated clients) for the group in 20.0.3 takes only 50ms.

I tested this with Keycloak running in a docker container and a local installation of a postgres database on my pc.

I know performance tests are very tricky. But the difference between these systems is significant enough that I'm making an issue out of it.

### Version

20.0.3

### Expected behavior

Getting a token in this case should not last that long.

### Actual behavior

Getting a token takes significant amount of time.

### How to Reproduce?

perftest.zip](https://github.com/keycloak/keycloak/files/10684154/perftest.zip)

In the attached ZIP file is everything you need to reproduce this.

Run `docker build -t keycloakperf .` to create an optimized image for postgres database (that is used in the docker-compose.yml).

Edit the docker-compose.yml to set the used database. You will need a postgres database somewhere to test with. Run `docker compose up` to start keycloak with the previously built image.

Open the JMeter project (BigData.jmx) in JMeter and run it. This will create a new realm (perf) with one user (user001) in one group (group001) and the clients (client1 to client400). The composite roles of the clients will be assigned to the group. That takes a couple of minutes for 400 clients. However, you can set the number of created clients in the project.

When all is finished get a token for the created user with
`curl --location --request POST 'http://localhost:8080/realms/perf/protocol/openid-connect/token' 
--header 'Content-Type: application/x-www-form-urlencoded' 
--data-urlencode 'username=user001' 
--data-urlencode 'password=userpw' 
--data-urlencode 'client_id=admin-cli' 
--data-urlencode 'grant_type=password'`

As you can see the client_id in the login request is admin-cli and not one of the created clients, so the resulting access_token has no resource access entries.

To test with keycloak 18.0.0 just change the used version in the Dockerfile, build the image with that and run the same test with a fresh database.

**Comment by pedroigor** [comment]
@jhollmannk That is what I call a reproducer. Really helpful and nice. Checking it out.

**Comment by pedroigor** [comment]
@mrpatrickpilch Run the reproducer using versions:

* `upstream`/`nightly`
* `20.0.3`
* `18.0.0`

I did not see a great difference across these versions. All of them have a response time between 2.5s and 3s, using default cache settings. I'm using loca caches.

```
<local-cache name="realms">
            <encoding>
                <key media-type="application/x-java-object"/>
                <value media-type="application/x-java-object"/>
            </encoding>
            <memory max-count="1000000"/>
        </local-cache>

         <local-cache name="users">
            <encoding>
                <key media-type="application/x-java-object"/>
                <value media-type="application/x-java-object"/>
            </encoding>
            <memory max-count="1000000"/>
        </local-cache>
```

I'm not running tests using a container, but using a standalone dist. The options I'm using to start the server are:

```
./kc.sh start --http-enabled=true --hostname-strict=false --cache=local --db postgres --db-username <user> --db-password <password> --metrics-enabled true --health-enabled true --db-url-database=<database> --cache-config-file=cache-local.xml
```

For `18.0.0`, you need to set the `--auto-build` option too.

Once the cache is hot, the `curl` request you provided finishes around 100ms for all of them. One thing I noticed tough, is that time is a bit worse for `20.0.3` but not close to what you mentioned (6500ms).

@keycloak/performance Would you be able to give us a third eye on this and give us a more "independent" evaluation of the result?

**Comment by pedroigor** [comment]
@jhollmannk Btw, one thing I noticed though is that with `18.0.0` I can not even access `http://localhost:8080/admin/master/console/` because the page keeps loading. Are you experiencing the same?

**Comment by kami619** [comment]
@pedroigor sure thing, we are taking a look into this too

**Comment by jhollmannk** [comment]
Oh dear! I didn't check the results of the REST calls of the JMeter project. I use the admin user of the master realm to set up the data. As that takes a few minutes I set the access token lifespan to 30 minutes. And that REST call does not work in version 18.0.0 (bad request). You can see that in JMeter in the "View Results Tree" node of the "Prepare Settings" or "Create data" thread group.

So in version 18 the data is not completely there as after 1 minute the token of the admin user expires. That is why version 18 was so fast in comparison.

Now I corrected the access token lifespan REST call and the versions now show similar timings. I'm attaching the corrected JMeter project here in case you want to try this again.

[BigData.zip](https://github.com/keycloak/keycloak/files/10766535/BigData.zip)

Version 20 is in fact a bit faster.

Still the login calls are around 7s in version 20 and around 9s in version 18. That is pretty consistent even on different PCs of my colleagues. Running Keycloak as a container or directly in a shell does not make a difference.

And yes I also cannot access the admin console in version 18. Somewhere along the load up the application tries to switch to `https://localhost:8443` to load `step1.html`. I guess the problem does not occur when running as a container as I set proxy=edge and set hostname and hostname-port (even if in this case Keycloak is not running behind a proxy).

Odd that you didn't see any difference between 18 and 20 as the test data was not completely there for version 18.

I will try to set the cache settings in my test environment. Maybe that will speed up the process. I have never done this before, so I will need some time for that.

I guess that makes this issue more or less invalid as the assumed difference between the versions is not really there. I apologize for the inconvenience.

Could you point me to some log setting maybe to see where all the time is lost on our test systems? Maybe we can pinpoint some location where the time is spent.

**Comment by pedroigor** [comment]
@jhollmannk I'll try it again. We do have our own benchmark too, see https://github.com/keycloak/keycloak-benchmark. 

It is a great tool that @kami619 and @ahus1 are working with. It provides you with both provisioning and benchmarking tools. Perhaps it needs some tweaks to fit into your test scenario but it is worthy to try it out.

I appreciate your time on this and we're always open to get feedback from the community about benchmarks. There is always something to learn/improve.

> Could you point me to some log setting maybe to see where all the time is lost on our test systems? Maybe we can pinpoint some location where the time is spent.

Do you mean for the admin console issue on 18?

**Comment by jhollmannk** [comment]
I tried a different cache setting and it had no effect until I read the description for `--cache-config-file` more thoroughly and saw that this path is relative to the `conf` folder. That did the trick. Phew! I was a bit worried there.

I was able to debug the token endpoint and can see why the cache is important in this case. The roles for the groups the user is in are collected and then the composite roles are resolved. Now that can be a lot of work and takes some time.
Only after that the intersection between that and the roles of the client for which the login is made is calculated. In this case the intersection is empty.

Thank you so much for your effort! Otherwise I would not have realized the error I made with version 18.0.0 while setting up the data.

You should never compare absolute durations of a performance test between different systems. You had durations of 2,5s to 3s, we had 5s-7s. That might just be our machines.

The cache works as intended and we will set this to sensible values depending on the data load.

As soon as I have time I want to look into the benchmark project. It would be nice to have this run on windows with RancherDesktop as the Kubernetes provider. Maybe I can manage that. Thanks for pointig out that project.

As there was nothing wrong with Keycloak at the first place I'm closing this issue.
